### PR TITLE
Feature/query cxx nick e aircft

### LINKED ISSUE #521: Write/test yaml converting AIRCFT bufr file to IODA
Description

[ ] Write a yaml that can convert the bufr file. Use the mnemonics determined in #520 
[ ] Use the yaml to convert into an IODA-formatted netCDF file.
[ ] Add the outputs to the testoutput directory.
[ ] Add tests to the CMakeLists.txt files
[ ] Have the tests pass.
Requirements

All important mnemonics need to be known. #520  will need to be finished.
Acceptance Criteria (Definition of Done)

It will be finished when all files have been checked, pushed, and tests have passed.
Dependencies

The bufr2ioda.x converter update must be finished.

### PR DESCRIPTION
**Pull request by nicholasesposito**
## Description
Adds a number of yamls that convert aircft data (both bufr and prepbufr) to IODA-formatted netcdf.

The following files were added:
```
test/testinput/gdas.t12z.acft_profiles.prepbufr
test/testinput/gdas.t12z.aircft.tm00.bufr_d
test/testinput/gdas.t12z.aircar.tm00.bufr_d

test/testinput/bufr_ncep_prepbufr_aircft.yaml
test/testinput/bufr_ncep_aircft.yaml
test/testinput/bufr_ncep_aircar.yaml

test/testoutput/gdas.t12z.acft_profiles.prepbufr.nc
test/testoutput/gdas.t12z.aircft.tm00.nc
test/testoutput/gdas.t12z.aircar.tm00.nc

```
The following file was edited:
`CMakeLists.txt`

The following tests were added/edited:
```
test_iodaconv_bufr_aircft2ioda
iodaconv_validate_testoutput_gdas.t12z.aircft.tm00.nc      

test_iodaconv_bufr_aircar2ioda
iodaconv_validate_testoutput_gdas.t12z.aircar.tm00.nc

test_iodaconv_prepbufr_aircft2ioda
iodaconv_validate_testoutput_gdas.t12z.acft_profiles.prepbufr.nc   
```

The ObsSpace.yaml from the following branch was used to validate the data.
`feature/query_cxx_NE_aircft`

### Issue(s) addressed

fixes #521 #523 #747 #748 #749 #750 #751 #752 #753 #754 #755 #756 

The issue containing links to the issue that showed the comparison between GSI and AIRCFT prepbufr are here: #758 

## Acceptance Criteria (Definition of Done)

PR is finished when all yamls are accepted.

## Dependencies

None

## Impact
None

### CODE REVIEW

File: test/testinput/bufr_ncep_prepbufr_aircft.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_prepbufr_aircft.yaml:193]: @nicholasesposito,  units: "m"

File: test/testinput/bufr_ncep_prepbufr_aircft.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_prepbufr_aircft.yaml:235]: units: "K"

**Comment by YoulongXia-NOAA** [comment]
@nicholasesposito, I checked all units for your ioda output, Greg used abbreviation but you use full names. I think you should follow Greg's conventions by using abbreviations such as "kilogram kilogram-1" into "kg kg-1".  could you please make the corresponding modifications? I will take the second round review after your changes.

**Comment by nicholasesposito** [comment]
@YoulongXia-NOAA   I think they used to be that way which is why I didn't  change them. But fair enough, I'll make those changes.

**Comment by YoulongXia-NOAA** [comment]
@nicholasesposito, can you see my all comments  as I jsu add single comment but did not do "submit review" action.

**Comment by nicholasesposito** [comment]
Recommended changes have been made for units @YoulongXia-NOAA

File: test/testinput/bufr_ncep_aircft_nc004001.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_aircft_nc004001.yaml:134]: units: "deg"

File: test/testinput/bufr_ncep_aircft_nc004002.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_aircft_nc004002.yaml:142]: units: "deg"

File: test/testinput/bufr_ncep_aircft_nc004003.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_aircft_nc004003.yaml:195]: units: "deg"

File: test/testinput/bufr_ncep_aircft_nc004006.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_aircft_nc004006.yaml:190]: units: "deg"

File: test/testinput/bufr_ncep_aircft_nc004009.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_aircft_nc004009.yaml:197]: units: "deg"

File: test/testinput/bufr_ncep_aircft_nc004010.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_aircft_nc004010.yaml:260]: units: "deg"

File: test/testinput/bufr_ncep_aircft_nc004011.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_aircft_nc004011.yaml:144]: units: "degree"

File: test/testinput/bufr_ncep_aircft_nc004103.yaml
Comment (YoulongXia-NOAA) [inline test/testinput/bufr_ncep_aircft_nc004103.yaml:197]: units: "deg"

**Review by YoulongXia-NOAA — CHANGES_REQUESTED** [review]
@nicholasesposito , I did not pick this issue up before. After I go through the yaml file again and double check Greg's conventions, I found it. Whe your time is available,  please fix it.

**Review by rmclaren — COMMENTED** [review]
There are modifications to the GNSSRO and the aircraft YAML files in this one pull request... Is that intentional?

Please see the comments I had for pull request https://github.com/JCSDA-internal/ioda-converters/pull/904. They apply to this pull request as well.

More general question (for everyone especially @emilyhcliu:) ). Why is it that sometimes we combine data for different subsets in a BUFR file to make one big data set, and other times we seem to want to create a totally seperate output file for each subset?

**Comment by nicholasesposito** [comment]
Recommended changes made and CMakeLists.txt updated!
@rmclaren The modifications to GNSSRO are not intentional. Not sure why it keeps doing that. I'm gonna try to work that out tomorrow

**Comment by nicholasesposito** [comment]
Nevermind, I figured it out. All that's left is the aircft files

**Comment by nicholasesposito** [comment]
The issue containing links to the issue that showed the comparison between GSI and AIRCFT prepbufr are here: #758

**Comment by nicholasesposito** [comment]
@YoulongXia-NOAA the ObsSpace.yaml has the unit for Wind Direction as "degree", so I'm changing it back to that.

**Comment by nicholasesposito** [comment]
Aircraft files have been updated. All tests I added pass.

  Test #518: test_iodaconv_bufr_aircar2ioda
  Test #587: iodaconv_validate_testoutput_gdas.aircar.t00z.20210801.nc
  Test #597: iodaconv_validate_testoutput_gdas.t12z.aircar_bufr_thinned.tm00.nc
  Test #517: test_iodaconv_bufr_aircft2ioda
  Test #519: test_iodaconv_prepbufr_aircft2ioda
  Test #598: iodaconv_validate_testoutput_gdas.t12z.aircft_bufr_thinned.tm00.nc

**Comment by YoulongXia-NOAA** [comment]
> @YoulongXia-NOAA the ObsSpace.yaml has the unit for Wind Direction as "degree", so I'm changing it back to that.

Sure, just be consistent with one.  Greg's doc has also some inconsistency but I am not sure which we need to follow. The ObsSpace.yaml may be a good way. Thank you @nicholasesposito

**Review by rmclaren — APPROVED** [review]

**Review by YoulongXia-NOAA — APPROVED** [review]

**Comment by srherbener** [comment]
The build itself is failing. The compilers are throwing errors which can be viewed on CDash.

**Comment by rmclaren** [comment]
It doesn't look like the right version of NCEPLIB-bufr is beign used. I created the following ticket to get them to create a numbered tag for us that include the latest software changes that we need.

https://github.com/NOAA-EMC/NCEPLIBS-bufr/issues/206

**Comment by rmclaren** [comment]
I created the following ticket to get them to create a numbered tag for us that include the latest software changes that we need.

https://github.com/NOAA-EMC/NCEPLIBS-bufr/issues/206

**Comment by srherbener** [comment]
> I created the following ticket to get them to create a numbered tag for us that include the latest software changes that we need.
> 
> [NOAA-EMC/NCEPLIBS-bufr#206](https://github.com/NOAA-EMC/NCEPLIBS-bufr/issues/206)

@rmclaren thanks for helping with this! Hopefully, we'll get the tag soon.

**Comment by nicholasesposito** [comment]
Like the adpsfc, the input file was originally missing each subtype. Per Emily's request. I used cat on each thinned file, created a new one, added it, and added the new output.

**Comment by nicholasesposito** [comment]
Going through this at the moment and making updates similar to what we did for adpsfc. Will comment here when the changes are done.

**Review by rmclaren — APPROVED** [review]

**Comment by nicholasesposito** [comment]
Perfect. Thanks everyone.

**Comment by nicholasesposito** [comment]
@emilyhcliu It's basically done. Trying to add QualityMarker/percentConfidence to the ObsSpace.yaml but I'm getting a warning about it. 
 Warning: Attribute 'units' is present but is not listed as a required or optional attribute in the spec.
The units in the nc file match the ObsSpace.yaml ("1", since it's a percent). 

I wonder if it has to do with the fact that it's a Quality Marker? I think normally Quality Markers usually don't have units so that may be the issue. But in the bufr table [website ](https://www.nco.ncep.noaa.gov/sib/jeff/bufrtab_tableb.html) PCCF is under the quality marker section. What do you think we should do about this?

**Comment by nicholasesposito** [comment]
Checking the conventions document, I realized that other "percent confidence" variables were MetaData. So I changed my percent confidence by MetaData, and updated my percentConfidence entry in my new ObsSpace.yaml to require MetaData too. I updated the files and all tests now pass.

**Comment by emilyhcliu** [comment]
@nicholasesposito @rmclaren I am reviewing this now.

**Comment by emilyhcliu** [comment]
Validation for Winds (IODA vs. GSI)
gdas t12z acft_profiles_windEastward_dim0
gsi_ncep_prepbufr_aircft_uwind
gdas t12z acft_profiles_windNorthward_dim0
gsi_ncep_prepbufr_aircft_vwind

**Comment by emilyhcliu** [comment]
Validation for Specific Humidity (IODA vs. GSI)
gdas t12z acft_profiles_specificHumidity_dim0
gsi_ncep_prepbufr_acft_sphum

**Comment by nicholasesposito** [comment]
@emilyhcliu ok I will check it out.

**Comment by emilyhcliu** [comment]
@nicholasesposito I just verified, it is ACRN 

For stationIdentation: 

use RPID for dump 001, 002, 003 
use ACRN for dump 004, 006, 009, 010, 011, 103 

Please also make sure the mnemonics in the bufr dump can be associated with mnemonics in the prepbufr.  We went through the similar processing for ADPSFC.  That's why we organized the YAML in terms of groups for both bufr and prepbufr.

I will continue reviewing the dump YAMLs.    

Thanks.

**Comment by emilyhcliu** [comment]
@nicholasesposito 
We missed one thing in the prepbufr.
We need to add "ACID" (aircraft flight number) in the prepbufr

**Comment by emilyhcliu** [comment]
@nicholasesposito I found a few things that need to be modify in prepbufr YAML.
Please see the inline review comment.

**Comment by nicholasesposito** [comment]
@rmclaren @emilyhcliu 
I am trying to add the mnemonic "ACID" to the AIRCFT Prepbufr yaml/output. When I put the mnemonic in, the yaml doesn't like it. I believe the reason is due to this: the input file actually contains two dataypes in it (AIRCFT and AIRCAR). With the exception of one mnemonic (ACID), the two sequences are exactly the same. All other variables are fine.

From the bufr table:
```
| AIRCAR   | HEADR  ACID  {PRSLEVLA}  <LATCORSQ>  <LONCORSQ>                 
| AIRCFT   | HEADR  <ACID_SEQ>  {PRSLEVLA}  <LATCORSQ>  <LONCORSQ> 
```
where
```
| ACID_SEQ | ACID  
```

The queries for ACID, according to print_queries.x are: ``` AIRCFT/ACID_SEQ/ACID``` and  ```AIRCAR/ACID``` and the query for prepbufrDataLevelCategory is  ``` */PRSLEVLA/CAT  ```
Regardless of what combination I use for my query, I get the failure as seen at the bottom of this message. Some individual ones do work, however...

The following queries do NOT work:
```
[*/ACID, */ACID_SEQ/ACID]
AIRCFT/ACID_SEQ/ACID*/ACID_SEQ/ACID
```

But the following queries DO work:
```
*/ACID
AIRCAR/ACID
```

Initially we used prepbufrDataLevelCategory as our group_by field. I'm thinking that this is the issue has to do with ACID having two different queries. What do you suggest we do in this situation? When I try to group_by: aircraftFlightNumber (ACID), I get a SegFault.

The working directory is here:
/scratch1/NCEPDEV/da/Nicholas.Esposito/JEDI/AIRCFT_PB_CONV_2020110112/

The test yaml is:
acid_problem.yaml

The input file is:
testinput/gdas.t12z.acft_profiles.prepbufr

The queries:
testinput/gdas.t12z.prepbufr.acft_profiles.queries

The bufr table is:
testinput/acft_profile_table


The error I get when I run acid_problem.yaml is:
```
[Nicholas.Esposito@hfe02 AIRCFT_PB_CONV_2020110112]$ ../bufr2ioda.x acid_problem.yaml
Start
Executing Queries
Building Bufr Data
Exporting Data
Finished [0.023s]
HDF5-DIAG: Error detected in HDF5 (1.12.0) thread 0:
  #000: ../../src/H5D.c line 280 in H5Dopen2(): name parameter cannot be an empty string
    major: Invalid arguments to routine
    minor: Bad value
terminate called after throwing an instance of 'std::_Nested_exception<ioda::Exception>'
  what():       Reason: An exception occurred inside ioda while opening a variable.
        name:
        source_column:  0
        source_filename:        /scratch1/NCEPDEV/da/Nicholas.Esposito/JEDI/ioda-bundle_20220630/ioda-bundle/ioda/src/engines/ioda/src/ioda/Has_Variables.cpp
        source_function:        virtual ioda::Variable ioda::detail::Has_Variables_Base::open(const string&) const
        source_line:    108

Aborted
```

**Comment by nicholasesposito** [comment]
It was discovered that we do not need the AIRCFT/ACID_SEQ/ACID because GSI does not use it. It only uses */ACID. The AIRCFT ACID sequence was removed and the conversion works without it.

The changes have been pushed!

**Comment by emilyhcliu** [comment]
> @rmclaren @emilyhcliu I am trying to add the mnemonic "ACID" to the AIRCFT Prepbufr yaml/output. When I put the mnemonic in, the yaml doesn't like it. I believe the reason is due to this: the input file actually contains two dataypes in it (AIRCFT and AIRCAR). With the exception of one mnemonic (ACID), the two sequences are exactly the same. All other variables are fine.
> 
> From the bufr table:
> 
> ```
> | AIRCAR   | HEADR  ACID  {PRSLEVLA}  <LATCORSQ>  <LONCORSQ>                 
> | AIRCFT   | HEADR  <ACID_SEQ>  {PRSLEVLA}  <LATCORSQ>  <LONCORSQ> 
> ```
> 
> where
> 
> ```
> | ACID_SEQ | ACID  
> ```
> 
> The queries for ACID, according to print_queries.x are: ` AIRCFT/ACID_SEQ/ACID` and  `AIRCAR/ACID` and the query for prepbufrDataLevelCategory is `*/PRSLEVLA/CAT ` Regardless of what combination I use for my query, I get the failure as seen at the bottom of this message. Some individual ones do work, however...
> 
> The following queries do NOT work:
> 
> ```
> [*/ACID, */ACID_SEQ/ACID]
> AIRCFT/ACID_SEQ/ACID*/ACID_SEQ/ACID
> ```
> 
> But the following queries DO work:
> 
> ```
> */ACID
> AIRCAR/ACID
> ```
> 
> Initially we used prepbufrDataLevelCategory as our group_by field. I'm thinking that this is the issue has to do with ACID having two different queries. What do you suggest we do in this situation? When I try to group_by: aircraftFlightNumber (ACID), I get a SegFault.
> 
> The working directory is here: /scratch1/NCEPDEV/da/Nicholas.Esposito/JEDI/AIRCFT_PB_CONV_2020110112/
> 
> The test yaml is: acid_problem.yaml
> 
> The input file is: testinput/gdas.t12z.acft_profiles.prepbufr
> 
> The queries: testinput/gdas.t12z.prepbufr.acft_profiles.queries
> 
> The bufr table is: testinput/acft_profile_table
> 
> The error I get when I run acid_problem.yaml is:
> 
> ```
> [Nicholas.Esposito@hfe02 AIRCFT_PB_CONV_2020110112]$ ../bufr2ioda.x acid_problem.yaml
> Start
> Executing Queries
> Building Bufr Data
> Exporting Data
> Finished [0.023s]
> HDF5-DIAG: Error detected in HDF5 (1.12.0) thread 0:
>   #000: ../../src/H5D.c line 280 in H5Dopen2(): name parameter cannot be an empty string
>     major: Invalid arguments to routine
>     minor: Bad value
> terminate called after throwing an instance of 'std::_Nested_exception<ioda::Exception>'
>   what():       Reason: An exception occurred inside ioda while opening a variable.
>         name:
>         source_column:  0
>         source_filename:        /scratch1/NCEPDEV/da/Nicholas.Esposito/JEDI/ioda-bundle_20220630/ioda-bundle/ioda/src/engines/ioda/src/ioda/Has_Variables.cpp
>         source_function:        virtual ioda::Variable ioda::detail::Has_Variables_Base::open(const string&) const
>         source_line:    108
> 
> Aborted
> ```

@nicholasesposito You found a bug.  This is good.  Please open an issue using the content of this specific comment.     Thanks.

**Comment by nicholasesposito** [comment]
Will do @emilyhcliu

**Comment by emilyhcliu** [comment]
@nicholasesposito Looking into the YAML for aircraft dump file.
One thing to note is that the ObsValue pressure (PRLC) for both AIRCAR and AIRCFT are all missing.
However, the pressure (POB) in the prepbufr file has valid pressure values in it.  So, PRLC should be calculated from height (ZOB) with standard atmosphere assumption.  Need to document how the pressures are calculated.  We have similar cases in ADPSFC.  For some subset, the surface pressure data are missing and therefore, calculated.

File: test/testinput/bufr_ncep_aircar.yaml
Comment (nicholasesposito) [reply test/testinput/bufr_ncep_aircar.yaml:45]: Remove for now. Save it in the master file.

File: test/testinput/bufr_ncep_prepbufr_aircft.yaml
Comment (nicholasesposito) [reply test/testinput/bufr_ncep_prepbufr_aircft.yaml:72]: In GSI, but not used.

File: test/testinput/bufr_ncep_prepbufr_aircft.yaml
Comment (nicholasesposito) [reply test/testinput/bufr_ncep_prepbufr_aircft.yaml:84]: In GSI but not used.

**Comment by nicholasesposito** [comment]
@emilyhcliu 
I plotted the full data of my netcdf file using the software plotting that I use, and I got different results than you. In our discussion, it seemed like the South America and Japan data was not being shown and is in primarily 4103. Using the full data, I got pull map worth of imagery, including the 4103 data.

I plotted temperature using the script here. I can plot more when I get back next week.
/scratch1/NCEPDEV/da/Nicholas.Esposito/JEDI/python_plots/qu_aircft/plot_ioda_aircft_query_Tair.py

Can you send me your plots so I can further look at them, too? Thanks.
fulldata_Tair

**Comment by nicholasesposito** [comment]
AMDAR and noAMDAR have been added and the original aircft bufr dump yaml has been taken out.

**Review by emilyhcliu — APPROVED** [review]

**Comment by emilyhcliu** [comment]
@rmclaren Please go ahead and merge this PR.  Thanks.
After you merge this one into feature/query_cxx.  I am going to approve the PR for feature/query_cxx.

**Comment by srherbener** [comment]
> @rmclaren Please go ahead and merge this PR. Thanks. After you merge this one into feature/query_cxx. I am going to approve the PR for feature/query_cxx.

Please don't merge this until `feature/query_cxx` is merged into develop. There are too many changes here that would be added without the awareness of the reviewers who have already approved PR #1006.

Thanks!

**Comment by emilyhcliu** [comment]
> > @rmclaren Please go ahead and merge this PR. Thanks. After you merge this one into feature/query_cxx. I am going to approve the PR for feature/query_cxx.
> 
> Please don't merge this until `feature/query_cxx` is merged into develop. There are too many changes here that would be added without the awareness of the reviewers who have already approved PR #1006.
> 
> Thanks!
@srherbener I see.  We will hold this one.

**Comment by BenjaminRuston** [comment]
@nicholasesposito this PR is dead now should we close this one and just focus on the new one

**Comment by BenjaminRuston** [comment]
@nicholasesposito  we can close this as it is superseded by PR#1071

https://github.com/JCSDA-internal/ioda-converters/pull/1071

correct?

**Comment by nicholasesposito** [comment]
Yes I'll close it

**Comment by nicholasesposito** [comment]
This PR has been superseded by PR #1071 and is now closed.

**Pull request closed without merging**
